seb_openedx/middleware.py

```diff
@@ -1,27 +1,130 @@
-            return render_to_response('seb-403.html', status=403)
+                    access_denied = False
+
+        if access_denied:
+            is_courseware_view = bool(view_func.__name__ == courseware.views.index.CoursewareIndex.__name__)
+            if is_courseware_view:
+                return self.courseware_error_response(request, *view_args, **view_kwargs)
+            return self.generic_error_response(request, course_key)
+        return None
+
+    def is_whitelisted_view(self, request, course_key):
+        """ First broad filter: whitelisting of paths/tabs """
+
+        # Whitelisting logic by alias
+        aliases = {
+            'discussion.views': 'discussion',
+            'course_wiki.views': 'wiki',
+            'openedx.features.course_experience': 'course-outline',
+        }
+
+        views_module = inspect.getmodule(request.resolver_match.func).__name__
+
+        alias_current_path = next(sixfilter(lambda k: aliases[k] if views_module.startswith(k) else None, aliases), None)
+
+        if alias_current_path in SEB_WHITELIST_PATHS:
+            return True
+
+        # Whitelisting xblocks when courseware
+        if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):
+            return True
+
+        # Whitelisting by url name
+        if request.resolver_match.url_name:
+            url_names_allowed = list(SEB_WHITELIST_PATHS) + ['jump_to', 'jump_to_id']
+            for url_name in url_names_allowed:
+                if request.resolver_match.url_name.startswith(url_name):
+                    return True
+
+        return False
+
+    def is_blacklisted_chapter(self, request, course_key):
+        """ Second more granular filter: blacklisting of specific chapters """
+        chapter = request.resolver_match.kwargs.get('chapter')
+
+        if not SEB_BLACKLIST_CHAPTERS:
+            return False
+
+        if chapter in SEB_BLACKLIST_CHAPTERS:
+            return True
+
+        if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):
+            usage_id = request.resolver_match.kwargs.get('usage_id')
+            if usage_id:
+                chapter = get_chapter_from_location(usage_id, course_key)
+                if chapter in SEB_BLACKLIST_CHAPTERS:
+                    return True
+        return False
+
+    def courseware_error_response(self, request, *view_args, **view_kwargs):
+        """ error response when a chapter is being blocked """
+        html = Fragment()
+        html.add_content(render_to_string('seb-403-error-message.html', {}))
+        SebCoursewareIndex.set_context_fragment(html)
+        return SebCoursewareIndex.as_view()(request, *view_args, **view_kwargs)
+
+    def generic_error_response(self, request, course_key):
+        """ generic error response, full page 403 error (with course menu) """
+        courseware = get_courseware_module()
+        try:
+            course = courseware.courses.get_course(course_key, depth=2)
```

Comment:
Would it make sense to wrapp `get_course` instead of `get_courseware_module` ?

Reply:
We use get_courseware_module on another file as well: seb_courseware_index.py
And we probably gonna be using some attributes (and deeper ones) of the courseware_module in the near future, so having separate backends for each function will result in too many files.